src/node/flows.ml

```diff
@@ -206,14 +209,21 @@ let try_to_commit_state_hash ~prev_validators state block signatures =
         ~block_payload_hash:block.payload_hash
         ~withdrawal_handles_hash:block.withdrawal_handles_hash
         ~state_hash:block.state_root_hash ~validators ~signatures)
+let hash_new_state_root state protocol update_state =
+  let task_pool = !get_task_pool () in
+  let snapshot_ref, snapshots =
+    Snapshots.add_snapshot_ref ~block_height:protocol.block_height
+      state.Node.snapshots in
+  let _task : unit Task.promise =
+    Task.async task_pool (fun () ->
+        let hash, data = Protocol.hash protocol in
+        Snapshots.set_snapshot_ref snapshot_ref { hash; data }) in
+  update_state { state with snapshots }
+
 let rec try_to_apply_block state update_state block =
   let%assert () =
     ( `Block_not_signed_enough_to_apply,
       Block_pool.is_signed ~hash:block.Block.hash state.Node.block_pool ) in
-  let%assert () =
-    ( `Invalid_state_root_hash,
-      block_matches_current_state_root_hash state block
-      || block_matches_next_state_root_hash state block ) in
```

Comment:
Sorry, why is this removed again? Shouldn't we always have the next state root hash? And if it's late it should hash it synchronously IIRC.

Reply:
If it's late, it does not start hashing asynchronously - this PR is fully async. 

The way it currently works is that if you miss a state root hash, we apply it anyway (because this check is removed), but you won't sign blocks. In #250, we'll add a way to sign those blocks if you later finish the state root hash. 

If we add a synchronous hashing mechanism, then there's no need to remove this check, and also no need for #250, but it may cause nodes to hash the same state on multiple cores. I think we could implement something more complex like "if you don't have the hash you need but it's being hashed on another core, wait until it's finished" - I'd have to think about it though.

Reply:
The check can be removed but we should add a comment here, this should make you go out of sync so that you don't sign blocks and then if you finish hashing it late and it matches you can go back to being in sync(as your hypothesis was correct).

Reply:
added a comment in latest commit